internal/consensus/reactor.go

```diff
@@ -835,24 +833,42 @@ func (r *Reactor) queryMaj23Routine(ctx context.Context, ps *PeerState) {
 	timer := time.NewTimer(0)
 	defer timer.Stop()
 
-OUTER_LOOP:
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	signal := make(chan struct{})
```

Comment:
Given that we have a context channel, and we seem to wait on it only in selections where we are _also_ waiting on `ctx`, could we replace `closeSignal` with an explicit call to `cancel()` in those places? (It's safe to call a cancel function any number of times, it already includes the one-shot protection)

Reply:
So I think the context takes care of situations where we're doing shut down of the entire system, but there's one of these daemon threads per-peer and the peer so if the channel closes or there's a problem (which isn't propogated up to the enclosing loop/context,) I wanted the close signal from one of the worker threads to kill the other threads.) 

It's also potentially the case that we're more likely to hit lock contention on one of the calls *before* the channel sends, which means if someone encounters an error before a different thread starts trying to send, the context won't be cancled yet and we'll block/leak threads? 

It's probably overkill, but I think it's a bit more orderly.
It's probably overkill, but I think it's a bit more orderly.

Reply:
Sorry, maybe I'm being dense, but I don't understand the argument. The context is created right here, and in all the locations (I think) where the code is currently waiting on `signal` it's also waiting on `ctx`. 

> I wanted the close signal from one of the worker threads to kill the other threads

Did you mean _not_ kill the other threads? Because if we really want everything that shares a context to stop, then I'd expect the context _is_ the right signal to use? 

But if you mean we don't want to kill everything for one failure, then I can see that.

Not trying to be a pedant here, but I'm definitely still confused about the goal.

Reply:
If any of the channel sends (which are running in worker go routines started in an iteration of this loop encounter an error (context close or otherwise) there's no way to have _this iteration_ exit early if one of the sends fails. 

The context could be live, but there would be an error, and without the signal thread, if any of the worker threads encounter an error the context doesn't get canceled. I did write this before I added an extra context cancellation, which perhaps makes this redundant.